### PR TITLE
Add test leg to the PR build to run libraries tests on iOS x64 Simulators

### PR DESCRIPTION
**Pull request by premun**
Forked from https://github.com/dotnet/runtime/pull/37057

Adds the iOS test leg that will execute selected BCL tests on an iOS x64 Simulator (13.5) using the `OSX.1015.Amd64.Open` queue which is our on-prem Mac Mini infrastructure (60 machines).

To prepare the Helix job, we are using the Arcade Helix SDK, more specifically the [XHarness Helix SDK](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/Readme.md) which uses the [XHarness tool](https://github.com/dotnet/xharness) to control the test run.

### CODE REVIEW

File: src/libraries/sendtohelix.proj
Comment (premun) [inline src/libraries/sendtohelix.proj:110]: **Help wanted:**
Can someone please let me know if there's a better way to find all directories ending with `.app`?
Or should I just hardcode the path inside the test assets artifact that contains the folders?

The trailing `/` will be fixed in Helix SDK later

Comment (directhex) [reply src/libraries/sendtohelix.proj:111]: Suggestions from @akoeplinger on internal chat:

should be clearer with `<XHarnessAppFolderToTest Include="@(_appBundleDirs)" Condition="%(Identity.EndsWith('.app')) or %(Identity.EndsWith('.app/'))` I think
assuming these directories are at the edge of the tree, rather than somewhere in the middle
or even easier: `<XHarnessAppFolderToTest Include="$([System.IO.Directory]::GetDirectories('$(TestArchiveTestsRoot)', '*.app', System.IO.SearchOption.AllDirectories))" />`

File: global.json
Comment (premun) [inline global.json:18]: Can we try this version?

```suggestion
    "Microsoft.DotNet.Helix.Sdk": "5.0.0-beta.20308.1",
```

File: src/libraries/sendtohelix.proj
Comment (premun) [inline src/libraries/sendtohelix.proj:46]:
```suggestion
    <MicrosoftDotNetXHarnessCLIVersion>1.0.0-prerelease.20308.1</MicrosoftDotNetXHarnessCLIVersion>
```

File: src/libraries/sendtohelix.proj
Comment (premun) [inline src/libraries/sendtohelix.proj:44]: This ArchiveTests doesn't do anything I think. It was my attempt that didn't work and then you resolved it in yaml

File: src/libraries/sendtohelix.proj
Comment (premun) [inline src/libraries/sendtohelix.proj:107]:
```suggestion
      <!-- Find all directories named *.app -->
```

File: eng/pipelines/libraries/build-job.yml
Comment (premun) [inline eng/pipelines/libraries/build-job.yml:111]: This will need to be made dynamic so that it works for other platforms as well but I am not sure how it works and how it gets consumed

More information here: https://github.com/dotnet/runtime/commit/2b7852b3d039b7d40f2623a3ec54e7d23eb93790

Comment (ViktorHofer) [reply eng/pipelines/libraries/build-job.yml:111]: Opened https://github.com/dotnet/runtime/pull/37652 to unblock you.

File: src/libraries/sendtohelix.proj
Comment (safern) [inline src/libraries/sendtohelix.proj:56]: Why is this change needed? `BuildAllConfigurations` has nothing to do with iOS.

File: src/libraries/sendtohelix.proj
Comment (safern) [inline src/libraries/sendtohelix.proj:45]: Use the property in Versions.props instead?

File: src/libraries/sendtohelix.proj
Comment (premun) [reply src/libraries/sendtohelix.proj:56]: You're right, it's not

File: src/libraries/sendtohelix.proj
Comment (premun) [reply src/libraries/sendtohelix.proj:45]: There was none, but I added it.

File: src/libraries/tests.proj
Comment (premun) [inline src/libraries/tests.proj:22]: @safern won't this influence other platforms as well? Maybe we want to add a special clause for `TargetsMobile`?

**Comment by directhex** [comment]
Should have been reading the errors in more detail

Looks like most of our errors right now are from Helix:

```
[07:22:51] dbug: test[0]
      07:22:51.3221030 4382 Execution timed out after 30 seconds and the process was killed.
[07:22:51] dbug: test[0]
      07:22:51.3269460 Process exited with 137
[07:22:51] dbug: test[0]
      07:22:51.3281860 mlaunch failed when listing simulators but trying to parse the results anyway
[07:22:51] dbug: test[0]
      07:22:51.3352210 Failed to find/create simulator (attempt 1/3):
      System.Exception: Failed to list simulators - no XML with devices found. mlaunch timed out with 137)
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.LoadDevices(ILog log, Boolean includeLocked, Boolean forceRefresh, Boolean listExtraData) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 80
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.FindOrCreateDevicesAsync(ILog log, String runtime, String devicetype, Boolean force) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 192
         at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.SimulatorLoader.FindSimulators(TestTargetOs target, ILog log, Boolean createIfNeeded, Boolean minVersion) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:line 385
         at Microsoft.DotNet.XHarness.iOS.AppRunner.RunApp(AppBundleInformation appInformation, TestTargetOs target, TimeSpan timeout, TimeSpan testLaunchTimeout, String deviceName, String companionDeviceName, Boolean ensureCleanSimulatorState, Int32 verbosity, XmlResultJargon xmlResultJargon, String[] skippedMethods, String[] skippedTestClasses, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.iOS/AppRunner.cs:line 115
[07:22:51] dbug: test[0]
      07:22:51.3401080 /Applications/Xcode115.app/Contents/Developer/usr/bin/simctl create "iPhone-X (iOS-13-5) - created by XHarness" com.apple.CoreSimulator.SimDeviceType.iPhone-X com.apple.CoreSimulator.SimRuntime.iOS-13-5
```

**Comment by premun** [comment]
@directhex I noticed this issue a while ago and logged it here: https://github.com/dotnet/xharness/issues/283

mlaunch fails to get a simulator so I added a retry (notice the `attempt 1/3` which mostly solved the symptom but as I currently don't have any time dedicated to XHarness I didn't solve the underlying issue.

I will try to get some time as this will be a blocker soon and I need to investigate if it's machine-specific issue

File: eng/testing/AppleRunnerTemplate.sh
Comment (premun) [inline eng/testing/AppleRunnerTemplate.sh:46]:
```suggestion

# Restart the simulator to make sure it is tied to the right user session
xcode_version=11.4
xcode_path="/Applications/Xcode${xcode_version/./}.app"
simulator_app="$xcode_path/Contents/Developer/Applications/Simulator.app"
sudo pkill -9 -f "$simulator_app"
open -a "$simulator_app"
```

File: eng/testing/AppleRunnerTemplate.sh
Comment (premun) [inline eng/testing/AppleRunnerTemplate.sh:57]:
```suggestion
echo "XHarness artifacts: `ls -lh $XHARNESS_OUT`"
```

File: eng/testing/AppleRunnerTemplate.sh
Comment (premun) [inline eng/testing/AppleRunnerTemplate.sh:49]: Here I would also recommend specifying `--timeout` and `--launch-timeout`.
The former is for the overall run, the latter for how long it takes to boot the simulator, install the app and start it.

The failures you are seeing (mlaunch fails to find a simulator) have a retry logic and can take time from what I've seen so before we fix it, it could help increasing the timeouts (say 15 minutes for the launch timeout and 30-40 for the overall timeout)

**Review by premun — COMMENTED** [review]
@directhex I noticed this issue a while ago and logged it here: https://github.com/dotnet/xharness/issues/283

mlaunch fails to get a simulator so I added a retry (notice the `attempt 1/3` which mostly solved the symptom but as I currently don't have any time dedicated to XHarness I didn't solve the underlying issue.

I will try to get some time as this will be a blocker soon and I need to investigate if it's machine-specific issue

**Comment by directhex** [comment]
@premun could this be something as stupid as not giving mlaunch enough time to finish?

Here's a good run:

```
[05:25:47] dbug: test[0]
      05:25:47.1394590 /private/tmp/helix/working/A9DE099B/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.20352.3/tools/netcoreapp3.1/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app --listsim=/tmp/tmpmlrVXC.tmp --output-format=XML
[05:26:08] dbug: test[0]
      05:26:08.8570810 Simulator configuration written to /tmp/tmpmlrVXC.tmp
[05:26:08] dbug: test[0]
      05:26:08.9515020 Process exited with 0
[05:26:09] dbug: test[0]
      05:26:08.9536870 Simulator listing returned:
```

That took 22 seconds to return any results

Here's a bad one:

```
[02:06:49] dbug: test[0]
      02:06:49.7723290 /private/tmp/helix/working/A9DE099B/p/microsoft.dotnet.xharness.cli/1.0.0-prerelease.20352.3/tools/netcoreapp3.1/any/../../../runtimes/any/native/mlaunch/bin/mlaunch --sdkroot /Applications/Xcode115.app --listsim=/tmp/tmp3dqcCa.tmp --output-format=XML
[02:07:19] dbug: test[0]
      02:07:19.8645850 Process 4094 didn't exit within 00:00:30 and will be killed
```

Timeout is 30 seconds

How close to the wire is it getting?

**Comment by premun** [comment]
@directhex I have observed these behaviours so far:
- mlaunch takes about 1 second to list simulators successfully
- mlaunch takes really long and times out
- mlaunch takes really long and times out **but** lists simulators successfully
- mlaunch returns a non-0 exit code
- mlaunch returns a non-0 exit code **but** lists simulators successfully

What I did is implement a retry logic to give 3 attempts to this and also even when mlaunch fails, I try to read the XML files with the results so sometimes mlaunch fails and I still get the list of simulators.

I haven't had much time investigating this though so in order to unblock us, I will need to talk to my manager first.

In the meantime I did some experimentation on one of the machines where I see a lot of failures but it is quite hard to reproduce sometimes...

**Comment by premun** [comment]
I have an attempt to fix the issue: https://github.com/dotnet/xharness/pull/288
So we will see..

**Comment by premun** [comment]
@directhex I have fixed the mlaunch issues and we can move forward with this if you wish.

I rebased and we can see how the tests go.

After we merge the newest Helix SDK (after an Arcade rollout), the `<LaunchTimeout>` value will also come into effect and improve edge cases where simulators would take long to boot. However, I don't expect this to happen after the fix I made to the mlaunch problem.

**Comment by premun** [comment]
So it seems the issue is still on-going, will revisit tomorrow.

**Comment by directhex** [comment]
Just doing one sanity build, to make sure it wasn't a fluke that it passed last time, but given last time was a glorious pass, I am feeling optimistic

**Comment by premun** [comment]
@directhex I have run about 20 builds and the **Send to Helix** part was 100% fine. The Build step however failed occasionally. It would take 2 hours and then timeout. The build seemed to keep on building, just slowly.

I have also observed this on my local Mac Mini - building the iOS apps took waaay longer than I have seen in AzDO.

**Comment by directhex** [comment]
@premun I've actually just been handed a bug about build timeouts on a completely unrelated job - https://github.com/dotnet/runtime/issues/40933

Infra issue?

**Comment by directhex** [comment]
Wait, can it be infra if it's happening on both Mac and Linux? Like, Mac is hosted hardware, Linux is Azure VMs, they couldn't be more different yet both are showing this strange slowdown behaviour. New CoreCLR bug?

**Comment by premun** [comment]
@directhex I was observing this already few months ago. No idea whether it's tied to a specific machine too since re-runs usually help.

**Comment by directhex** [comment]
Hm, seems to legit break the Wasm lane

**Comment by premun** [comment]
> Hm, seems to legit break the Wasm lane

@directhex I will have a look

**Comment by directhex** [comment]
@premun I can repro the issue locally.

**Comment by directhex** [comment]
Resolved.

File: eng/testing/AppleRunnerTemplate.sh
Comment (akoeplinger) [inline eng/testing/AppleRunnerTemplate.sh:41]: hardcoding the Xcode version will make it difficult to keep this updated when new Xcode versions are rolled out, or locally where I just have Xcode.app.

why do we not use `/Applications/Xcode.app`?

Comment (premun) [reply eng/testing/AppleRunnerTemplate.sh:41]: We have several versions installed side-by-side on our on-prem machines. It is a practice adopted from Xamarin as it is then super clear on which version is used.

You could change this script to get the version from `xcode-select -p`

Comment (akoeplinger) [reply eng/testing/AppleRunnerTemplate.sh:41]: Yes using `xcode-select` would be much better.

Comment (premun) [reply eng/testing/AppleRunnerTemplate.sh:41]: I will give it a go, and also will stop specifying the iOS version since if we install a new Xcode, it might not have 13.5 available

Comment (premun) [reply eng/testing/AppleRunnerTemplate.sh:41]: We basically won't be able to target one version of iOS with confidence.. We can also not specify the version and XHarness will grab the latest

File: eng/testing/AppleRunnerTemplate.sh
Comment (akoeplinger) [inline eng/testing/AppleRunnerTemplate.sh:49]: I feel like I commented about this in some PR already but I don't like the `ls` here. It means I can't cmd+click the path in my terminal to open the folder anymore since the path is no longer printed.

Comment (premun) [reply eng/testing/AppleRunnerTemplate.sh:49]: Changed

**Review by akoeplinger — COMMENTED** [review]
LGTM!

**Review by akoeplinger — APPROVED** [review]

**Comment by directhex** [comment]
https://helix.dot.net/api/2019-06-17/jobs/ba701e37-fa40-4a78-92d9-729043b36d9c/workitems/Microsoft.Extensions.Configuration.Json.Tests/console

Timed out installing the test app?

File: eng/testing/AppleRunnerTemplate.sh
Comment (safern) [inline eng/testing/AppleRunnerTemplate.sh:41]: Will we support filtering tests or any xunit parameters? Like `-notrait Category=Outerloop`? If so, this is wrong and should use `RunScriptCommand` in the MSBuild side to set this command.

File: src/libraries/sendtohelixhelp.proj
Comment (safern) [inline src/libraries/sendtohelixhelp.proj:30]: 30 minutes? That is too much, not? Should we set this special timeout for iOS only? We used to have 15 mins timeout.

File: src/libraries/sendtohelixhelp.proj
Comment (safern) [inline src/libraries/sendtohelixhelp.proj:31]: Ditto.

File: eng/testing/AppleRunnerTemplate.sh
Comment (akoeplinger) [reply eng/testing/AppleRunnerTemplate.sh:41]: Ideally yes, but we first need a way to flow arbitrary command line arguments to the iOS/Android test apps and that will need some changes in xharness and the iOS/Android test runners, I don't think we need to block the PR on that.

Comment (safern) [reply eng/testing/AppleRunnerTemplate.sh:41]: Sounds good. We just need to be conscious that PR validation is running outerloop tests for iOS as well, which might be causing the slow runs.

Comment (akoeplinger) [reply eng/testing/AppleRunnerTemplate.sh:41]: ah no, we added handling for that already with https://github.com/dotnet/runtime/blob/3c1b48ee29f13cac93a86829374b917375a49a2d/eng/testing/tests.mobile.targets#L76 which is read by the testrunner so Outerloop and failing tests should already be excluded.

Comment (safern) [reply eng/testing/AppleRunnerTemplate.sh:41]: Ok, sounds good. Maybe we should do it how we do it in other places passing them as an argument but that is in the future, no need to block this PR.

**Comment by premun** [comment]
> https://helix.dot.net/api/2019-06-17/jobs/ba701e37-fa40-4a78-92d9-729043b36d9c/workitems/Microsoft.Extensions.Configuration.Json.Tests/console
> 
> Timed out installing the test app?

Looks like it happened again but on the very same machine (**dci-macpro-06**). I will investigate what's up

File: src/libraries/sendtohelixhelp.proj
Comment (premun) [reply src/libraries/sendtohelixhelp.proj]: I reset the previous ones and set one specifically for iOS/tvOS. I, however, don't understand what the `Scenario` is for and whether I should also have one for non-empty `Scenario` and iOS/tvOS?

Comment (safern) [reply src/libraries/sendtohelixhelp.proj]: Nope. Scenarios are for special test modes that are translated into a job that sets an environment variable before running the tests.

At the moment there is no such thing for iOS/tvOS, so let's just set it for empty scenario and iOS/tvOS. If the need comes later on we can adjust.

**Comment by premun** [comment]
I took the problematic machine offline for investigations so let's see whether the rest goes green now.
I will have time to investigate next week.

**Comment by premun** [comment]
Alright, looks everything is green. So I wonder if we want to check this in or not. The reason is that sometimes the iOS build times out when it is on one of the slow agents (or whatever the reason is), so this will bring flakey-ness. On the other hand, I am not sure whether this is not happening already then?

**Comment by safern** [comment]
I think if we don't check this in we won't find those flaky issues and fix them. However, the question I have is, do we have the capacity yet to enable this on all PRs and CI?

**Comment by premun** [comment]
The capacity for the iOS tests itself is OK. The tests only take 6 minutes right now. We have 59 machines for that and there's 155 of work items so the test itself is usually super short. The capacity for building the iOS tests is coming from AzDO and I am not sure how are we doing there. I will try to find out.

**Comment by safern** [comment]
Building the product and tests took 36 mins in last build which is not bad.

**Comment by directhex** [comment]
Practically speaking, what blockers remain on this PR?

**Comment by premun** [comment]
> @directhex Practically speaking, what blockers remain on this PR?

So, from my point of view (and please consider that this is my first and only interaction with `dotnet/runtime` so far and I don't have experience in how the overall PR builds behave):
- Building the iOS apps times out every now and then when a "slow agent" is processing the build. A re-run is needed. I am not sure whether this causes problems with other legs already since it doesn't seem to be tied with this particular leg but..
- I don't know how often it happens on regular basis that you need to re-run a leg in a PR build so I don't know what is the "regular flakey-ness".
- This week one machine somehow went bad and installation of apps started to hang. I took it offline for investigations but I am not sure about the root cause still and I am not sure this won't happen again to bring more flakey-ness. (but as @safern noted, maybe the only way to find out is to check it in and observe).
- I am not sure how to approach the question of capacity. I personally think we are fine for iOS but again - maybe the only way to see is to check it in.

So.. I don't know how flakey the PR builds in `dotnet/runtime` are already and whether we won't bring a lot of pain by checking this in, but I guess this is up to you guys. I don't know whether we shouldn't just flag the leg as not required and observe first or what is the regular procedure when adding new platforms, but I just don't have experience with this repository to decide this.

Other than that, the iOS tests part seems quite stable and fast, so..

**Comment by safern** [comment]
> Building the iOS apps times out every now and then when a "slow agent" is processing the build. A re-run is needed. I am not sure whether this causes problems with other legs already since it doesn't seem to be tied with this particular leg but..

We're already seeing this in other OSX, iOS legs and it seems to be an AzDo bug which seems to be under investigation so I wouldn't be concerned by this and also seems like it has somewhat being mitigated.

> This week one machine somehow went bad and installation of apps started to hang. I took it offline for investigations but I am not sure about the root cause still and I am not sure this won't happen again to bring more flakey-ness. (but as @safern noted, maybe the only way to find out is to check it in and observe).

I still stand my point the best way to mitigate issues is to check-in and investigate issues that come show up.

> I am not sure how to approach the question of capacity. I personally think we are fine for iOS but again - maybe the only way to see is to check it in.

Given the number of machines you shared above and the time it is taking to process work items it seems fine. We won't know this with certainty until we check it in, and as we enable more test projects and run on more PRs if we hit an issue we can always disable the leg as we did for a couple of days for WASM until we got more capacity. 

Overall, this LGTM and we should try checking it in, specially since master is already 6.0.

**Review by safern — APPROVED** [review]

**Comment by steveisok** [comment]
@premun Please give your approval on the PR and we can merge.

File: eng/testing/AppleRunnerTemplate.sh
Comment (ViktorHofer) [inline eng/testing/AppleRunnerTemplate.sh:4]: I'm trying to get rid of the runcommands templating mechanism as having individual runner scripts for different targets (ios/android/wasm) seems to be sufficient. Doing that in https://github.com/dotnet/runtime/pull/39923. That will basically undo this change here. Any objections?

Comment (safern) [reply eng/testing/AppleRunnerTemplate.sh:4]: No, I support that change as we should be able to define it via MSBuild.

Comment (steveisok) [reply eng/testing/AppleRunnerTemplate.sh:4]: No objections from me.

**Pull request merged**
